### PR TITLE
ADBDEV-4904-27: Add check sendingChunkTransportState is not NULL

### PR DESCRIPTION
**Pull request by red1452**
Add check sendingChunkTransportState is not NULL

The sendingChunkTransportState pointer is initialized conditionally and may be
NULL. Add missing check in a place where sendingChunkTransportState->numConns is
dereferenced to be 0 if it's NULL.

Introduce numConns which contains sendingChunkTransportState->numConns or 0 to
avoid checking sendingChunkTransportState for NULL every iteration.

### CODE REVIEW

File: src/backend/cdb/motion/ic_tcp.c [hunk 1]
```diff
@@ -1775,7 +1775,7 @@ SetupTCPInterconnect(EState *estate)
 		i = 0;
 		while (n > 0 &&
 			   outgoing_count < expectedTotalOutgoing &&
-			   i < sendingChunkTransportState->numConns)
+			   i < sendingChunkTransportState ? sendingChunkTransportState->numConns : 0)
```
Comment (RekGRpth) [inline src/backend/cdb/motion/ic_tcp.c]: Pointer sendingChunkTransportState may be dereferenced [below](https://github.com/arenadata/gpdb/pull/768/files#diff-8de7e292e2ab3f0d6b1229dd48de95c2589b047830aab7f3777cf8de0ef24bf1R1780)

Comment (RekGRpth) [inline src/backend/cdb/motion/ic_tcp.c]: I'm not sure about the sequence of operations here. Maybe we should add explicit parentheses?

File: src/backend/cdb/motion/ic_tcp.c [hunk 2]
```diff
@@ -1462,9 +1463,9 @@ SetupTCPInterconnect(EState *estate)
 
 		/* Outgoing connections */
 		outgoing_count = 0;
-		n = sendingChunkTransportState ? sendingChunkTransportState->numConns : 0;
+		numConns = sendingChunkTransportState ? sendingChunkTransportState->numConns : 0;
```
Comment (bandetto) [inline src/backend/cdb/motion/ic_tcp.c:1466]: Can it really be `NULL`? `sendingChunkTransportState` is returned from:
https://github.com/arenadata/gpdb/blob/df6f5babee0643069a3f3bbe13f97be1d1c35746/src/backend/cdb/motion/ic_tcp.c#L1376
Where:
https://github.com/arenadata/gpdb/blob/df6f5babee0643069a3f3bbe13f97be1d1c35746/src/backend/cdb/motion/ic_tcp.c#L514-L529

Comment (RekGRpth) [reply src/backend/cdb/motion/ic_tcp.c:1466]: it is initialized conditionally
https://github.com/arenadata/gpdb/blob/df6f5babee0643069a3f3bbe13f97be1d1c35746/src/backend/cdb/motion/ic_tcp.c#L1375-L1376

Comment (bandetto) [reply src/backend/cdb/motion/ic_tcp.c:1466]: Is this part alright?
https://github.com/arenadata/gpdb/blob/df6f5babee0643069a3f3bbe13f97be1d1c35746/src/backend/cdb/motion/ic_tcp.c#L1383

Comment (RekGRpth) [reply src/backend/cdb/motion/ic_tcp.c:1466]: yes, expectedTotalOutgoing is set only when startOutgoingConnections

**Comment by bandetto** [comment]
May the description be updated to:
> sendingChunkTransportState pointer is initialized conditionally and may be NULL.
Add missing check in a place where sendingChunkTransportState->numConns is
dereferenced to be 0 if it's NULL.

> Introduce numConns which contains sendingChunkTransportState->numConns or 0 to
avoid checking sendingChunkTransportState for NULL every iteration.

**Comment by RekGRpth** [comment]
> May the description be updated to:
> 
> > sendingChunkTransportState pointer is initialized conditionally and may be NULL.
> > Add missing check in a place where sendingChunkTransportState->numConns is
> > dereferenced to be 0 if it's NULL.
> 
> > Introduce numConns which contains sendingChunkTransportState->numConns or 0 to
> > avoid checking sendingChunkTransportState for NULL every iteration.

updated

**Review by bandetto — APPROVED** [review]

**Pull request merged**